keras_nlp/src/models/segment_anything/segment_anything.py

```diff
@@ -0,0 +1,327 @@
+            input_image_size=prompt_encoder_input_image_size,
+            mask_in_channels=prompt_encoder_mask_in_channels,
+        )
+        # Prostprocessing layer
+        self.mask_decoder = SAMMaskDecoder(
+            transformer_dim=mask_decoder_transformer_dim,
+            transformer=self.transformer,
+            num_multimask_outputs=mask_decoder_num_multimask_outputs,
+            iou_head_depth=mask_decoder_iou_head_depth,
+            iou_head_hidden_dim=mask_decoder_iou_head_hidden_dim,
+        )
+        # === Functional Model ===
+        # Get the image encoder input -- Images
+        backbone_input = backbone.input
+        inputs = {
+            "images": backbone_input,
+            "points": keras.Input(shape=[None, 2], name="points"),
+            "labels": keras.Input(shape=[None], name="labels"),
+            "boxes": keras.Input(shape=[None, 2, 2], name="boxes"),
+            "masks": keras.Input(shape=[None, None, None, 1], name="masks"),
+        }
+
+        # Preprocess input prompts
+        prompt_embeddings = self.prompt_encoder(inputs)
+
+        # Define the mask decoder inputs
+        mask_decoder_inputs = {
+            "image_embeddings": backbone.output,
+            "image_pe": prompt_embeddings["dense_positional_embeddings"],
+            "sparse_prompt_embeddings": prompt_embeddings["sparse_embeddings"],
+            "dense_prompt_embeddings": prompt_embeddings["dense_embeddings"],
+        }
+
+        # Build the mask decoder
+        outputs = self.mask_decoder(mask_decoder_inputs)
+        super().__init__(inputs=inputs, outputs=outputs, **kwargs)
+
+        # === Config ===
+        self.transformer_num_layers = transformer_num_layers
+        self.transformer_hidden_size = transformer_hidden_size
+        self.transformer_intermediate_dim = transformer_intermediate_dim
+        self.transformer_num_heads = transformer_num_heads
+        self.prompt_encoder_hidden_size = prompt_encoder_hidden_size
+        self.prompt_encoder_image_embedding_size = (
+            prompt_encoder_image_embedding_size
+        )
+        self.prompt_encoder_input_image_size = prompt_encoder_input_image_size
+        self.prompt_encoder_mask_in_channels = prompt_encoder_mask_in_channels
+        self.mask_decoder_transformer_dim = mask_decoder_transformer_dim
+        self.mask_decoder_num_multimask_outputs = (
+            mask_decoder_num_multimask_outputs
+        )
+        self.mask_decoder_iou_head_depth = mask_decoder_iou_head_depth
+        self.mask_decoder_iou_head_hidden_dim = mask_decoder_iou_head_hidden_dim
+
+    def predict_step(self, *args, **kwargs):
+        if len(args) == 2:
+            args = (args[0], _add_placeholder_prompts(args[-1]))
+        else:
+            args = (_add_placeholder_prompts(args[0]),)
+
+        return super().predict_step(*args, **kwargs)
+
+    def fit(self, *args, **kwargs):
+        raise NotImplementedError(
```

Comment:
Is this true in the CV version as well? What is hard about supporting fit?

Reply:
yes, fit is not supported because of optional input prompts. The functional model needs all inputs to be specified. For predict, placeholders are added for the missing prompt inputs before calling predict_step.

Reply:
Got it! Let's ship with this, but @fchollet actually did add support for optional functional model inputs a bit ago. We might be able to leverage those here.

Reply:
SG!